_posts/2018-12-16-simple-front-end.md

```diff
@@ -0,0 +1,96 @@
+* Client side rendering
+	* Little or no HTML is delivered to the browser. Rather, javscript creates all the elements on the page.
+
+#### Pros
+* Most frameworks provide powerful two-way data binding to template and update what's on the page at any time without reloading
+* Feels more like a traditional desktop application 
+* Static scripts can be hosted on a CDN and scale well
+#### Cons
+* May perform poorly on lower end devices due to being script driven
+* First load may be slow due to large script dependencies and entire app needing to be sent at once
+* Mix of versioning, technologies, and compatibility can result in complicated build processes with polyfilling, transpiling, and dependency resolution
+
+### Somewhere in the middle
+What we'll cover today.
+* (mostly) Synchronous
+* Client side rendering
+	* Static HTML is delivered to the browser, but a small script provides the data for the dynamic content
+#### Pros
+* Performs well across a range of devices
+* Static parts of pages can be cached and hosted on a CDN
+#### Cons
+* New or refreshed content is accessed by a page reload, but that shouldn't be as much of a problem as it is with server side rendering since most of the page is cached
+
+## Solution
+
+So what are we actually talking about? In short it's two (three) dependencies:
+### Bootstrap
+Bootstrap makes it easy to quickly set up the style and layout of a page. It's been around since 2011 and hasn't changed too radically since then, but it has grown an abundance of documentation, community support, and usage guides. If you were inclined you could swap this out with any other library that does styling and layout like Google's material design for web. 
+### jQuery 
+jQuery is a transitive dependency from bootstrap, we'll only really use it directly to DRY up parts of our pages like headers and footers. When it comes to web libraries that are widely used, stable, and not going anywhere there is nothing that comes close to jQuery.
+### Handlebars
+From its readme:
+> Handlebars.js is an extension to the [Mustache templating language](http://mustache.github.com/) created by Chris Wanstrath. Handlebars.js and Mustache are both logicless templating languages that keep the view and the code separated like we all know they should be.
+
+Since 2012 handlebars has been the go-to javascript templating library. We'll use it to populate the parts of our pages which contain dynamic content (and very importantly, it will take care of escaping data to prevent code injection).
+### fetch
+Goodbye XMLHttpRequest! 
+I know I said three dependencies, but fetch is built into the browser making it a very reliable choice and technically not a dependency.
+
+> note: fetch is not supported in internet explorer, so if you need to support legacy browsers you'll want to use [github's fetch polyfil](https://github.com/github/fetch)
```

Comment:
💯 good job on the compatibility reminder.

Reply:
Have to credit @yacklebeam for reminding me about that one.